### PR TITLE
SFTTrainer support

### PR DESCRIPTION
**Pull request by michaelbenayoun**
# What does this PR do?

This PR adds two classes:

- `NeuronSFTConfig`
- `NeuronSFTTrainer`

Both of these classes achieve the same goal as their `trl` counterpart.

### CODE REVIEW

File: optimum/neuron/trainers.py [hunk 1]
```diff
@@ -1465,3 +1504,444 @@ class Seq2SeqNeuronTrainer(AugmentTrainerForNeuronMixin, Seq2SeqTrainer):
...
+                if torch_dtype != "auto" and not isinstance(torch_dtype, torch.dtype):
+                    raise ValueError(
+                        f"Invalid `torch_dtype` passed to the SFTConfig. Expected a string with either `torch.dtype` or 'auto', but got {torch_dtype}."
+                    )
+                model_init_kwargs["torch_dtype"] = torch_dtype
+
+        if infinite is not None:
+            warnings.warn(
+                "The `infinite` argument is deprecated and will be removed in a future version of TRL. Use `TrainingArguments.max_steps` or `TrainingArguments.num_train_epochs` instead to control training length."
+            )
```
Comment (JingyaHuang) [inline optimum/neuron/trainers.py:1607]: I like this kind of warning, I usually just remove it to keep less code and avoid users from using deprecated args.

**Review by JingyaHuang — COMMENTED** [review]
Thanks for the PR, it looks good! Just don't we need to add trl to the setup.py, add NeuronSFTTrainer to the API doc, and perhaps if possible have minimal test?

File: optimum/neuron/trainers.py [hunk 2]
```diff
@@ -1465,3 +1504,444 @@ class Seq2SeqNeuronTrainer(AugmentTrainerForNeuronMixin, Seq2SeqTrainer):
...
+            "max_seq_length",
+            "dataset_num_proc",
+            "dataset_batch_size",
+            "neftune_noise_alpha",
+            "model_init_kwargs",
+            "dataset_kwargs",
+            "eval_packing",
+            "num_of_sequences",
+            "chars_per_token",
+        ],
```
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply optimum/neuron/trainers.py]: I removed the deprecated arguments and requires `trl` >= 0.10.0`.

File: optimum/neuron/trainers.py (hunk 1 above)
Comment (michaelbenayoun) [reply optimum/neuron/trainers.py]: Done as well.

**Comment by michaelbenayoun** [comment]
I did not add `trl` in the setup, as for `peft`, because it is not really required. It's is only required for a subset of the features.

File: optimum/neuron/trainers.py [hunk 3]
```diff
@@ -1465,3 +1503,345 @@ class Seq2SeqNeuronTrainer(AugmentTrainerForNeuronMixin, Seq2SeqTrainer):
...
+        if is_peft_available():
+            from peft import PeftConfig, prepare_model_for_kbit_training
+
+        if args is None:
+            output_dir = "tmp_trainer"
+            warnings.warn(f"No `SFTConfig` passed, using `output_dir={output_dir}`.")
+            args = NeuronSFTConfig(output_dir=output_dir)
+        elif args is not None and args.__class__.__name__ == "NeuronTrainingArguments":
+            args_as_dict = args.to_dict()
+            # Manually copy token values as TrainingArguments.to_dict() redacts them
```
Comment (dacorvo) [inline optimum/neuron/trainers.py:1546]: This comes from the original trl, but I have no idea what this means ...

File: optimum/neuron/trainers.py [hunk 4]
```diff
@@ -1465,3 +1503,345 @@ class Seq2SeqNeuronTrainer(AugmentTrainerForNeuronMixin, Seq2SeqTrainer):
     """
     Seq2SeqTrainer that is suited for performing training on AWS Tranium instances.
     """
+
+
+class NeuronSFTTrainer(AugmentTrainerForNeuronMixin, SFTTrainer):
```
Comment (dacorvo) [inline optimum/neuron/trainers.py:1508]: Maybe add a comment here indicating how this differs from the original (ie what are the neuron specifics).

**Review by dacorvo — COMMENTED** [review]
As @JingyaHuang said, it would ne nice to have some unit tests before integrating this in the demo, to speed up integration by identifying issues early on.

File: optimum/neuron/trainers.py (hunk 4 above)
Comment (michaelbenayoun) [reply optimum/neuron/trainers.py]: Done!

**Comment by michaelbenayoun** [comment]
I have added tests. They do not check anything but run a small training job both with packed and unpacked datasets. If the training job succeeds, the test pass, otherwise it fails.

File: tests/test_trainers.py [hunk 5]
```diff
@@ -396,3 +396,70 @@ def preprocess_function(examples):
...
+    def parallel_sizes(self, request):
+        return request.param
+
+    def _test_sft_trainer(self, parallel_sizes, tmpdir, packing):
+        _, tp_size, pp_size = parallel_sizes
+
+        output_dir = Path(tmpdir)
+
+        dataset = load_dataset("databricks/databricks-dolly-15k", split="train")
+        # dataset = dataset.select(range(1000))
```
Comment (dacorvo) [inline tests/test_trainers.py:417]: nit: remove commented line

File: tests/test_trainers.py [hunk 6]
```diff
@@ -396,3 +396,70 @@ def preprocess_function(examples):
...
+        # Create Trainer instance
+        trainer = NeuronSFTTrainer(
+            model=model,
+            tokenizer=tokenizer,
+            train_dataset=dataset,
+            formatting_func=format_dolly,
+            args=sft_config,
+        )
+
+        trainer.train()
```
Comment (dacorvo) [inline tests/test_trainers.py:459]: Can't we verify that the loss goes down or something ?

File: optimum/neuron/trainers.py [hunk 7]
```diff
@@ -1455,13 +1497,368 @@ def save_state(self):
...
+
+
+class NeuronSFTTrainer(_TrainerForNeuron, _SFTTrainerTrainerInit):
+    """
+    `SFTTrainer` adapted for Neuron.
+
+    It differs from the original `SFTTrainer` by:
+        - Using `_TrainerForNeuron.__init__()` instead of `Trainer.__init__()`
+        - Using the `_TrainerForNeuron.train()` instead of `Trainer.train()`
+        - Adapts the `_prepare_non_packed_dataloader` to pad to max lenght. In the original `SFTTrainer` examples are
```
Comment (dacorvo) [inline optimum/neuron/trainers.py:1524]:
```suggestion
        - Adapts the `_prepare_non_packed_dataloader` to pad to max length. In the original `SFTTrainer` examples are
```

**Comment by dacorvo** [comment]
The new tests are failing:

```
FAILED tests/test_trainers.py::TestNeuronSFTTrainer::test_without_packing[dp=2] - TypeError: NeuronSFTConfig.__init__() got an unexpected keyword argument 'max_seq_length'
FAILED tests/test_trainers.py::TestNeuronSFTTrainer::test_with_packing[dp=2] - TypeError: NeuronSFTConfig.__init__() got an unexpected keyword argument 'max_seq_length'
FAILED tests/test_trainers.py::TestNeuronSFTTrainer::test_without_packing[tp=2] - TypeError: NeuronSFTConfig.__init__() got an unexpected keyword argument 'max_seq_length'
FAILED tests/test_trainers.py::TestNeuronSFTTrainer::test_with_packing[tp=2] - TypeError: NeuronSFTConfig.__init__() got an unexpected keyword argument 'max_seq_length'
```

File: tests/test_trainers.py (hunk 6 above)
Comment (michaelbenayoun) [reply tests/test_trainers.py:459]: It's a tiny random model. The `SFTTrainer` does not anything related to the loss anyways. It's just `Trainer` with dataset preparation abilities.

**Review by dacorvo — APPROVED** [review]
Thank you for this pull-request !

**Pull request merged**